src/libraries/System.Linq/src/System/Linq/Sum.cs

```diff
@@ -49,6 +72,101 @@ public static partial class Enumerable
             return sum;
         }
 
+        private static T SumSignedIntegersVectorized<T>(ReadOnlySpan<T> span)
+            where T : struct, IBinaryInteger<T>
+        {
+            Debug.Assert(span.Length >= Vector<T>.Count * 4);
+
+            ref T ptr = ref MemoryMarshal.GetReference(span);
+            int length = span.Length;
+
+            // Overflow testing for vectors is based on setting the sign bit of the overflowTracking
+            // vector for an element if the following are all true:
+            //   - The two elements being summed have the same sign bit. If one element is positive
+            //     and the other is negative then an overflow is not possible.
+            //   - The sign bit of the sum is not the same as the sign bit of the previous accumulator.
+            //     This indicates that the new sum wrapped around to the opposite sign.
+            //
+            // By bitwise or-ing the overflowTracking vector for each step we can save cycles by testing
+            // the sign bits less often. If any iteration has the sign bit set in any element it indicates
+            // there was an overflow.
+            //
+            // Note: The overflow checking in this algorithm is only correct for signed integers.
+            // If support is ever added for unsigned integers then the overflow check should be:
+            //   overflowTracking |= (accumulator & data) | Vector.AndNot(accumulator | data, sum);
+
+            Vector<T> accumulator = Vector<T>.Zero;
+
+            // Build a test vector with only the sign bit set in each element. JIT will fold this into a constant.
+            Vector<T> overflowTestVector = new(T.RotateRight(T.MultiplicativeIdentity, 1));
+
+            // Unroll the loop to sum 4 vectors per iteration
```

Comment:
What is the significance of unrolling to 4 vectors? Are spans of length between `Vector<T>.Count` and `Vector<T>.Count * 4` too short for vectorization to show tangible benefits, so we're leaving them out?

Reply:
Guessing it's related to overflow tracking requirements? My naive understanding of your algorithm is that you'd need to unroll to an even number vectors per iteration.

Reply:
Primarily it's for the performance gain, there isn't a requirement for an even number of vectors. The algorithm is a gain for as few as 2 vectors. Further down in the algorithm there is a loop that isn't unrolled that picks up trailing vectors or is the primary loop for arrays with 2 or 3 vectors.

Unrolling has two advantages. It allows us to swap the accumulator back and forth, which after JIT results in one fewer move operation per vector compared to the simple loop. It also should have the general advantages of unrolling such as fewer length check branches and better pipelining on the CPU. However, that's not my expertise so I haven't specifically tested for the latter. I just confirmed that x64 JIT looked right and the benchmarks were better with this loop unrolled.

Reply:
Correction, I misspoke, < 4 vectors isn't a performance gain in my testing. I tried the 2-3 vectors without unrolling but chose this approach based on benchmarks. I don't have the precise results anymore, but my recollection is that it was either a net loss or in some cases only a marginal gain. My assumption was that marginal gains weren't worth it since it could be an artifact of my i7 CPU and be a net loss on other hardware.

Reply:
Thanks